### DISCUSSION TITLE
How to perform parallel run in multiple nodes/machine in Jenkins

### DISCUSSION THREAD

**Discussion by bilarallen** [currents-dev/cypress-currents-reporter#81] [Q&A] [answered]
Hi! I hope someone can give an idea how we can perform parallel execution in multiple node/machine in Jenkins pipeline. I followed the instruction here https://currents.dev/readme/ci-setup/jenkins for setting up jenkins parallel execution. But it seems that it only execute in one machine.
![image](https://user-images.githubusercontent.com/65072341/225199506-b9e97842-fe4e-4dde-8604-ed70f5469702.png)
![image](https://user-images.githubusercontent.com/65072341/225199842-ae745048-3640-492d-abe3-6f9351c84bde.png)

Whenever I check the logs I see that every stages are being run in parallel but only in one machine
![image](https://user-images.githubusercontent.com/65072341/225200017-f7c37f6f-4c40-448e-a9a7-4dc6d69076ea.png)

My JenkinsPipeline is declarative and here is the setup
```
pipeline {
    agent {
        docker {
            label 'default'
            image 'cypress/base:14.16.0'
        
        }
    }
    stages {
        stage('Install Dependencies') { 
            steps {
                echo "Running build ${env.BUILD_ID} on ${env.JENKINS_URL}"
                sh 'npm ci'
                sh 'npm run cy:verify'
            }
        }
        
        stage('CBS UI Test') { 
            environment {
                CURRENTS_RECORD_KEY = credentials('currents-record-key')
                CURRENTS_PROJECT_ID = credentials('currents-master-projectid')
                CURRENTS_API_URL = credentials('currents-cloud-url')
                CYPRESS_trashAssetsBeforeRuns = 'false'
            }

            parallel {
                stage('Tester A') {
                    steps {
                        echo "Running build ${env.BUILD_ID} on Tester A"
                        sh "NO_COLOR=1 xvfb-run -a npx cypress-cloud run --parallel --record --key ${env.CURRENTS_RECORD_KEY} --ci-build-id ${env.GIT_BRANCH}-${env.BUILD_ID}"
                    }
                }

                stage('Tester B') {
                    steps {
                        echo "Running build ${env.BUILD_ID} on Tester B"
                        sh "NO_COLOR=1 xvfb-run -a npx cypress-cloud run --parallel --record --key ${env.CURRENTS_RECORD_KEY} --ci-build-id ${env.GIT_BRANCH}-${env.BUILD_ID}"
                    }
                }

                stage('Tester C') {
                    steps {
                        echo "Running build ${env.BUILD_ID} on Tester C"
                        sh "NO_COLOR=1 xvfb-run -a npx cypress-cloud run --parallel --record --key ${env.CURRENTS_RECORD_KEY} --ci-build-id ${env.GIT_BRANCH}-${env.BUILD_ID}"
                    }
                }

            }
           
        }
    }
}
```

Our built-in node has 6 executors and this is only the node where we can run our test as we specify by its label
![image](https://user-images.githubusercontent.com/65072341/225200554-97ceeb7e-89a8-4a5c-a6b2-8385d4e47081.png)

Our goal is to run it in parallel either in multiple node (where we will setup different machine) or in multiple executors using the same node.

Please take note that our Jenkins is dockerized that's why in my Jenkins pipeline I need to use docker cypress base image.

One thing to add is that even though our test is in parallel right now running our test in local is bit faster compare when running it in CI using Currents parallelization
![image](https://user-images.githubusercontent.com/65072341/225201283-40840007-1643-482c-b899-bddad2f0e2e9.png)
![image](https://user-images.githubusercontent.com/65072341/225201431-97d03078-9459-4a1f-8961-2bddaf690efd.png)

Hoping someone can help us in regards to this as this is the only factor left we consider before we finally buy a license from currents.dev 🤞 thanks!

**Comment by agoldis** [comment]
Hey, @bilarallen please send us a message via the in-app intercom, and we'll set up a session to check your situation. 

Looking at the Currents dashboard snapshot - there were 3 cypress runners involved in this run. So parallelization is working, not sure why the latest execution in Jenkins didn't activate all 3 stages. 

_re: our test in local is bit faster compared when running it in CI_

Your longest tests are running for between 4-7 minutes in CI vs only 2 minutes on the localhost. Not sure what type of hardware you're using, but that's the reason the difference in makespan is so significant. Also, keep in mind that recording tests require some extra work from the runner (capturing video for example)

**Reply by bilarallen** [reply]
Hi, @agoldis thanks for the response. I'll send a message in the intercom then 😄

**Comment by agoldis** [comment, marked as answer]
@bilarallen , to summarize:
- The parallelization is set up correctly
- The resource you're using is overloaded - the same spec takes between 3 to 7 minutes, I believe depending on the load. 

![currents-2023-03-14-23 39 30@2x](https://user-images.githubusercontent.com/1637928/225227357-caf904a5-9e67-48f7-95d9-457d39951a1b.png)

Try having a set of separate nodes for your cypress tests. Here is the [recommended spec](https://docs.cypress.io/guides/getting-started/installing-cypress#CPU)
